clean-modular-code/write-pseudocode.md

```diff
@@ -34,192 +39,298 @@ last_modified:
+    "processed_papers": processed_papers,
+    "mean_citations": mean_citations
+}
+```
+:::
 
 +++
 
-The above pseudocode could be passed to a llm (with more details about how you want it to create the code). 
+### You now have a cleaner, working script!
 
-You could also begin to translate it into Python code like this:
+By following the above steps, you now have a clean working workflow that you can 
+adapt and refactor even further. Let's pretend that you know you will need to 
+add two things to your workflow:
 
+* You know that your data is doing to be in a JSON file, rather 
+than provided to you as a Python dictionary. 
+* You also know that you will need to process multiple files. 
 
-1. Loop through and access each citation dictionary in the Python list. For each citation:  
-    1. Get the paper's publication date & turn it into a `datetime` object
-    1. Get the total citations the paper has
-    2. Add data to a pandas dataframe
-2. Calculate the mean citation number for all papers
+Have a look at the code below, consider using pseudocode to identify areas that you could 
+clean up even further to support a multi-file workflow.
 
 ```{code-cell} ipython3
 from datetime import datetime
 
-# Example list of papers with nested fields
-papers = [
-    {"title": ["Paper 1"], "pub_date": "2023/05/10", "citations": [5]},
-    {"title": ["Paper 2"], "pub_date": "2022/04/12", "citations": [3]},
-    {"title": ["Paper 3"], "pub_date": "2021/03/15", "citations": [8]},
+# Open the data in .JSON format rather than via an example 
+pubs = [
+    {"title": ["P1"], "pub_date": "2023/05/10", "citations": [5]},
+    {"title": ["P2"], "pub_date": "2022/04/12", "citations": [3]},
+    {"title": ["P3"], "pub_date": "2021/03/15", "citations": [8]},
 ]
-
-# Process papers in a loop
-for paper in papers:
-    if paper["citations"]:
-        pub_date = datetime.strptime(paper["pub_date"], "%Y/%m/%d")
-        avg_citations = sum(paper["citations"]) / len(paper["citations"])
-        print({
-            "title": paper["title"][0],  # Cleaning step
-            "pub_date": pub_date,        # Parsed datetime
-            "avg_citations": avg_citations
-        })
+# if there are multiple files, I will need a processing step for each file
+all_pubs = []
+for pub in pubs:
+    # Could this step be a function or multiple functions that cleans the data?
+    pub_date = datetime.strptime(pub["pub_date"], "%Y/%m/%d")
+    citation_count = pub["citations"][0]
+    all_pubs.append({"pub_date": pub_date, "citation_count": citation_count})
+
+all_pubs_df = pd.DataFrame(all_pubs)
+mean_citations = all_pubs_df["citation_count"].mean()
 ```
 
-You have now designed a workflow using pseudocode to process several sites worth of landsat data. 
++++ {"editable": true, "slideshow": {"slide_type": ""}}
 
-Of course, the pseudocode above is just the beginning. For each of the steps above, you need to flesh out how you can accomplish each task. 
+:::{todo}
```

Comment:
Next steps for the pseudocode lesson:

- [ ] flesh out the pseudocode further - have them turn a few steps into functions
- [ ] Potentially??? have them add multiple files which will make them think more about functions and processing steps